### DISCUSSION TITLE
SDXL Turbo released, 1 step SDXL

### DISCUSSION THREAD

**Discussion by saddam213** [saddam213/OnnxStack#52] [General]
https://huggingface.co/stabilityai/sdxl-turbo

Could be cool for Realtime if you have a decent GPU

**Update**
Looks like I can get 1 frame every 2 seconds using SDXL-Turbo at 1024x1024,  2 steps on a 3090 :)

**Comment by AshD** [comment]
I converted it to the Olive Onnx format
https://huggingface.co/softwareweaver/Sdxl-Turbo-Olive-Onnx

Regards,
Ash

**Comment by saddam213** [comment]
I seem to be getting just a black box on ImageToImage with the SDXL models, non olive models seems to work, but all the ones I have tried that are Olive just produced a black image, only if Inpaint or ImageToImage


![image](https://github.com/saddam213/OnnxStack/assets/4353684/2b04c3f9-2cbe-439e-9afc-e53c60f3635e)



1 frame every 2 seconds using SDXL-Turbo at 1024x1024, 2 steps on a 3090 :

**Reply by AshD** [reply]
Does it work with the Diffusers pipeline? And those extra parameters?

If not, I can open an issue with the Olive project

**Reply by saddam213** [reply]
Oh, I have never used Diffusers, ill get someone with python to give it a shot

**Reply by saddam213** [reply]
FYI: If I swap the Vae models for normal ones it works again

I wonder if we are overflowing F16 with 1024x1024 tensor math

**Reply by AshD** [reply]
So the issue is in the VaeEncoder and VaeDecoder optimized models?

**Reply by AshD** [reply]
Are there any new features for Tensor Math in Microsoft.ML 3.0 library that could help?

**Reply by saddam213** [reply]
Not sure, seems to be inside the Vae decoder, it could be an OnnxStack issue, so best to test outside the app before digging too deep

**Comment by AshD** [comment]
Opened an issue with Olive team. ORTStableDiffusionXLImg2ImgPipeline gives the same black image.
https://github.com/microsoft/Olive/issues/764

**Reply by saddam213** [reply]
Found this today by chance, could be related maybe?

https://huggingface.co/madebyollin/sdxl-vae-fp16-fix

**Reply by AshD** [reply]
Tried it. Same Black image with No errors :-(

I don't understand, if it is a VAE Decoder problem, how is TextToImage is working?

**Comment by AshD** [comment]
Hi @saddam213,  

So the guidance from the Olive team is to keep both the encoder/decoder FP32. 

I found the only the encoder needs to be FP32 for it to work and since you got Text to Image working, the decoder seems to be working fine.
I updated this model - https://huggingface.co/softwareweaver/stable-diffusion-xl-base-1.0-Olive-Onnx

Over the weekend, I will update all the other models.
Let me know if you see any issues.

Thanks,
Ash

**Reply by AshD** [reply]
All models are now updated.

**Reply by saddam213** [reply]
just tested `stable-diffusion-xl-base-1.0-Olive-Onnx` works perfect :)

**Comment by AshD** [comment]
The Turbo model should also work. Let me know the time it takes to generate
an image.

Regards,
Ash

**Comment by saddam213** [comment]
SDXL-Turbo
GPU: **3090**
VRAM: **15.8 GB**
4-Step: **1.9 seconds @512x512**

Works well for realtime

**Comment by saddam213** [comment]
LCM also have a SDXL model
[lcm-sdxl](https://huggingface.co/latent-consistency/lcm-sdxl)

If you have time to convert this one could be a good comparison

**Comment by AshD** [comment]
Thanks Adam. Benchmarks look great.

I will try to convert the lcm-sxl this week. I remember I had issues with
the SD LCM model. Might need the Olive team's help.

Regards,
Ash

**Comment by AshD** [comment]
Hi @saddam213 

I uploaded the LCM model here. https://huggingface.co/softwareweaver/Latent-Consistency-xl-Olive-Onnx

Can you verify that it works correctly. The SD LCM model had some issues. 

Thanks,
Ash

**Comment by AshD** [comment]
Tested with the Python code. Took 1.4 seconds (4090, steps=3, cfg=2.0) to generate the attached image.
prompt = "medieval castle by Caspar David Friedrich with a river running near it"


![test-dream](https://github.com/saddam213/OnnxStack/assets/233016/85d2f688-e021-4771-9b5b-b5ef398ee442)

**Comment by saddam213** [comment]
Sorry for the delay, just added the LatenetConsistancyXL pipeline :)

Model Works great
Took 2.6 seconds (3090, steps=3, cfg=2.0) to generate the attached image.
prompt = "medieval castle by Caspar David Friedrich with a river running near it"
![image](https://github.com/saddam213/OnnxStack/assets/4353684/3b2af5b4-e527-4d6e-bf11-0159125e68a0)

**Comment by AshD** [comment]
Cool. Thanks Adam.

This is a really awesome project 😀

I was looking for guidance to create onnx versions for MusicGen, SadTalker
and other models. If you know someone who could help there that would be
great.

Thanks,
Ash

**Reply by saddam213** [reply]
Unfortunately I am a model dumb, have no idea how to convert anything yet :/
